##  Name : Prajakta Pardeshi
###  CWID : A20352045
####  Homework Project: Crime Prediction - CS 584 Machine Learning
#### Expected Task : 1,2,3,4
#### Completed Task : 1,2,3,4,5,6.a

In [283]:
import numpy as np
import pandas as pd
from pandas.tools.plotting import scatter_matrix
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier,export_graphviz
import sklearn
from sklearn.preprocessing import LabelEncoder,LabelBinarizer
import os
import subprocess
from sklearn.ensemble import RandomForestClassifier
from sklearn import tree
from io import StringIO
from sklearn import metrics
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.model_selection import  cross_val_score
from sklearn.preprocessing import Imputer
from sklearn.metrics import precision_recall_fscore_support
from sklearn.naive_bayes import GaussianNB
from operator import itemgetter , attrgetter, methodcaller
from sklearn.metrics import average_precision_score
from sklearn.svm import LinearSVC
from sklearn.metrics import average_precision_score
from sklearn  import linear_model
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import RidgeCV
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model
from sklearn.linear_model import Perceptron
from sklearn.datasets import make_classification
from sklearn.ensemble import ExtraTreesClassifier

In [261]:
#Read data
dataset_final = pd.read_csv("E:\machine learning\project\Crime Prediction Data\communities-crime-clean.csv")
dataset = dataset_final.copy()
dataset['highcrime'] = dataset['ViolentCrimesPerPop'].map(lambda x: True if x > 0.1 else False)


## 1. Decision Trees

a.)	Create a new field “highCrime” which is true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise. What are the percentage of positive and negative instances in the dataset?

In [110]:
check_value=fil['highcrime'].value_counts()
true_value = (check_value[1]/(dataset['highcrime'].shape))*100
false_value =(check_value[0]/(dataset['highcrime'].shape))*100
print("The True percetage will be: %2f " %true_value)
print("The False percentage will be: %2f "%false_value)

The True percetage will be: 62.719518 
The False percentage will be: 37.280482 


What are the percentage of positive and negative instances in the dataset?


Answer : Hence, the  proportion of violent crimes per capita that is the total number of violent crimes per 100k population which has the  percentage greater than 0.1 is holding a higher percentage. Hence, we can justify that percentage of violent crime which is greater than 10% will be  considered highcrime area has a higher percentage of 62%.

b.	 Use DecisionTreeClassifier to learn a decision tree to predict highCrime on the entire dataset (remember to exclude the crime rate feature from the input feature set so you are not cheating). 

In [111]:
dataset1= dataset.copy()
y = dataset['highcrime']
dataset1 = dataset1.drop(['state','communityname','fold','ViolentCrimesPerPop','highcrime'], axis=1)


 i)What are the training accuracy, precision, and recall for this tree? 

In [113]:
clf_tree = DecisionTreeClassifier()
without_split_clf = clf_tree.fit(dataset1,dataset['highcrime'])
prediction = without_split_clf.predict(dataset1)

print("Accuracy: %2f" %clf_tree.score(dataset1,dataset['highcrime']))
print("The Recall is :",recall_score(y, prediction, average='weighted'))
print("The Precision is :",precision_score(y, prediction, average='weighted'))  
print ("\n\nThe report:-")
print (metrics.classification_report(y,prediction),"\n")

Accuracy: 1.000000
The Recall is : 1.0
The Precision is : 1.0


The report:-
             precision    recall  f1-score   support

      False       1.00      1.00      1.00       743
       True       1.00      1.00      1.00      1250

avg / total       1.00      1.00      1.00      1993
 



In the above result the precision, recall, accuracy all are 1 because we have trained and tested on the same dataset.

ii)What are the main features used for classification? Can you explain why they make sense (or not)?


In [266]:

my_features = dataset1.columns
imp_feature = clf_tree.feature_importances_
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature - {} ({})".format(i+1,my_features[indices[i]],imp_feature[indices[i]]))


Feature Ranking
1 feature - PctKids2Par (0.3586522413947397)
2 feature - racePctWhite (0.08881685598370975)
3 feature - racePctHisp (0.04858535565953847)
4 feature - PctLess9thGrade (0.02263011509341309)
5 feature - PctEmplManu (0.018823516048132382)
6 feature - MedRent (0.01558682991463624)
7 feature - PctHousOwnOcc (0.014165626021920983)
8 feature - OtherPerCap (0.013541829814025207)
9 feature - blackPerCap (0.013365762156192845)
10 feature - PctEmploy (0.013363970726593481)


The main features used for classification have been displayed above.
The features with the highest is the population of kids with two parents. This feature is given highest weight according to decision tree and looks reliable because the  kids with 2 parents are often prone to wrong direction and vulnerability.
The other features include percentage of population that is caucasian, percentage of population that is of hispanic heritage.

In [116]:
dataset1

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.01           0.61          0.21          0.83          0.02   
1           0.01           0.41          0.55          0.57          0.01   
2           0.03           0.34          0.86          0.30          0.04   
3           0.01           0.38          0.35          0.71          0.04   
4           0.04           0.37          0.32          0.70          0.21   
5           0.04           0.44          1.00          0.10          0.00   
6           0.41           0.37          1.00          0.02          0.03   
7           0.01           0.30          0.00          0.99          0.02   
8           0.00           0.39          0.31          0.75          0.02   
9           0.06           0.39          0.32          0.73          0.04   
10          0.07           0.41          0.53          0.57          0.05   
11          0.02           0.43          0.41          0.64          0.08   
12          0.01           0.46          0.67          0.47          0.02   
13          0.00           0.45          1.00          0.00          0.01   
14          0.04           0.33          0.33          0.73          0.02   
15          0.00           0.35          0.09          0.91          0.02   
16          0.05           0.32          0.55          0.56          0.04   
17          0.00           0.43          0.12          0.90          0.01   
18          0.02           0.23          0.16          0.85          0.08   
19          0.05           0.35          0.06          0.93          0.07   
20          0.01           0.44          0.26          0.80          0.01   
21          0.24           0.37          0.48          0.58          0.13   
22          0.00           0.54          0.34          0.71          0.07   
23          0.01           0.37          0.28          0.77          0.02   
24          0.01           0.36          0.15          0.83          0.16   
25          0.30           0.41          0.76          0.38          0.06   
26          0.28           0.45          0.83          0.34          0.04   
27          0.02           0.37          0.00          0.99          0.04   
28          0.01           0.44          0.44          0.64          0.05   
29          0.02           0.45          0.76          0.38          0.08   
...          ...            ...           ...           ...           ...   
1963        0.04           0.55          0.00          0.98          0.06   
1964        0.02           0.49          0.01          0.96          0.05   
1965        0.00           0.42          0.00          1.00          0.01   
1966        0.00           0.44          0.00          0.99          0.03   
1967        0.07           0.42          0.02          0.95          0.13   
1968        0.12           0.44          0.36          0.64          0.03   
1969        0.00           0.67          0.01          0.97          0.07   
1970        0.06           0.37          0.00          0.92          0.24   
1971        0.01           0.18          0.03          0.95          0.10   
1972        0.02           0.39          0.01          0.98          0.03   
1973        0.02           0.51          0.01          0.94          0.17   
1974        0.01           0.48          0.01          0.98          0.03   
1975        0.03           0.34          0.01          0.95          0.05   
1976        0.00           0.37          0.00          0.97          0.08   
1977        0.00           0.52          0.00          0.99          0.03   
1978        0.08           0.45          0.01          0.93          0.08   
1979        0.04           0.37          0.00          0.90          0.37   
1980        0.06           0.35          0.02          0.96          0.06   
1981        0.09           0.29          0.01          0.98          0.04   
1982        0.02           0.49          0.00          0.99          0.03   
19

c.	Now apply cross-validation (cross_val_score) to do 10-fold cross-validation to estimate the out-of-training accuracy of decision tree learning for this task.
        
        i.	What are the 10-fold cross-validation accuracy, precision, and recall?



In [123]:
print(" for folds")
accuracy_dt = cross_val_score(clf_tree , dataset1,y, cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_tree , dataset1, y,  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_tree , dataset1,  y, cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_tree , dataset1, y, cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_tree , dataset1, y, cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_tree , dataset1, y, cv=10, scoring='accuracy' ).mean())

 for folds
accuracy
[ 0.76        0.805       0.77        0.76884422  0.64824121  0.63316583
  0.72864322  0.70854271  0.79396985  0.70854271]
Precision
[ 0.74305556  0.848       0.75        0.79259259  0.7745098   0.73873874
  0.78195489  0.82692308  0.79285714  0.80357143]
Recall
[ 0.88   0.864  0.792  0.856  0.632  0.648  0.808  0.656  0.864  0.744]
Prec 0.779517
Recall 0.782400
Accuracy 0.728500


ii.	Why are they different from the results in the previous test?


Answer:  This is because while using cross-vaidation we are having different test and training dataset, whereas in first case we are testing on the same data we are training it. Thus, giving us accuracy, precison and recall all value as 1, whereas here we have a precision, Recall and precision value as 0.779, 0.782 and 0.7285 respectively a little lesser.


# 2. Linear Classification



 a.	Use GaussianNB to learn a Naive Bayes classifier to predict highCrime.

        i.	What is the 10-fold cross-validation accuracy, precision, and recall for this method?


In [124]:
clf_GB= GaussianNB()
print(" for folds")
accuracy_dt = cross_val_score(clf_GB , dataset1,dataset['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_GB, dataset1,dataset['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_GB, dataset1,dataset['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_GB , dataset1, dataset['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_GB , dataset1, dataset['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_GB , dataset1, dataset['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds
accuracy
[ 0.775       0.8         0.825       0.79899497  0.70351759  0.65326633
  0.81407035  0.73366834  0.71356784  0.79899497]
Precision
[ 0.86363636  0.92929293  0.95        0.92079208  0.94594595  0.86842105
  0.92307692  1.          0.77868852  0.93814433]
Recall
[ 0.76   0.736  0.76   0.744  0.56   0.528  0.768  0.576  0.76   0.728]
Prec 0.911800
Recall 0.692000
Accuracy 0.761608


What are the 10 most predictive features? This can be measured by the normalized absolute difference of means for the feature between the two classes:
|μ_T-μ_F |/(σ_T+σ_F )


In [125]:
dataset3 =  dataset1.copy()
clf_GB.fit(dataset3,dataset['highcrime'])
dataset3['predict'] = clf_GB.predict(fil1)
dataset3

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.01           0.61          0.21          0.83          0.02   
1           0.01           0.41          0.55          0.57          0.01   
2           0.03           0.34          0.86          0.30          0.04   
3           0.01           0.38          0.35          0.71          0.04   
4           0.04           0.37          0.32          0.70          0.21   
5           0.04           0.44          1.00          0.10          0.00   
6           0.41           0.37          1.00          0.02          0.03   
7           0.01           0.30          0.00          0.99          0.02   
8           0.00           0.39          0.31          0.75          0.02   
9           0.06           0.39          0.32          0.73          0.04   
10          0.07           0.41          0.53          0.57          0.05   
11          0.02           0.43          0.41          0.64          0.08   
12          0.01           0.46          0.67          0.47          0.02   
13          0.00           0.45          1.00          0.00          0.01   
14          0.04           0.33          0.33          0.73          0.02   
15          0.00           0.35          0.09          0.91          0.02   
16          0.05           0.32          0.55          0.56          0.04   
17          0.00           0.43          0.12          0.90          0.01   
18          0.02           0.23          0.16          0.85          0.08   
19          0.05           0.35          0.06          0.93          0.07   
20          0.01           0.44          0.26          0.80          0.01   
21          0.24           0.37          0.48          0.58          0.13   
22          0.00           0.54          0.34          0.71          0.07   
23          0.01           0.37          0.28          0.77          0.02   
24          0.01           0.36          0.15          0.83          0.16   
25          0.30           0.41          0.76          0.38          0.06   
26          0.28           0.45          0.83          0.34          0.04   
27          0.02           0.37          0.00          0.99          0.04   
28          0.01           0.44          0.44          0.64          0.05   
29          0.02           0.45          0.76          0.38          0.08   
...          ...            ...           ...           ...           ...   
1963        0.04           0.55          0.00          0.98          0.06   
1964        0.02           0.49          0.01          0.96          0.05   
1965        0.00           0.42          0.00          1.00          0.01   
1966        0.00           0.44          0.00          0.99          0.03   
1967        0.07           0.42          0.02          0.95          0.13   
1968        0.12           0.44          0.36          0.64          0.03   
1969        0.00           0.67          0.01          0.97          0.07   
1970        0.06           0.37          0.00          0.92          0.24   
1971        0.01           0.18          0.03          0.95          0.10   
1972        0.02           0.39          0.01          0.98          0.03   
1973        0.02           0.51          0.01          0.94          0.17   
1974        0.01           0.48          0.01          0.98          0.03   
1975        0.03           0.34          0.01          0.95          0.05   
1976        0.00           0.37          0.00          0.97          0.08   
1977        0.00           0.52          0.00          0.99          0.03   
1978        0.08           0.45          0.01          0.93          0.08   
1979        0.04           0.37          0.00          0.90          0.37   
1980        0.06           0.35          0.02          0.96          0.06   
1981        0.09           0.29          0.01          0.98          0.04   
1982        0.02           0.49          0.00          0.99          0.03   
19

   Printing top 10 features using NBgaussian

In [127]:
predict_true_dataset = dataset3[dataset3['predict'] == True]
predict_false_dataset = dataset3[dataset3['predict'] == False]


In [225]:
Nb_value = []
for i in range(len(fil1.columns)):
    Nb_value.append((abs(predict_true_dataset[[i]].mean() - predict_false_dataset[[i]].mean()) / predict_true_dataset[[i]].std() + predict_false_dataset[[i]].std() ))
sorted(Nb_value, key = itemgetter(0), reverse= True)[:10]

[PctKids2Par    1.881312
 dtype: float64, pctWInvInc    1.841672
 dtype: float64, PctPersOwnOccup    1.769371
 dtype: float64, medFamInc    1.763879
 dtype: float64, PctFam2Par    1.757952
 dtype: float64, medIncome    1.733898
 dtype: float64, FemalePctDiv    1.679538
 dtype: float64, PctYoungKids2Par    1.59169
 dtype: float64, PctHousLess3BR    1.579829
 dtype: float64, TotalPctDiv    1.575829
 dtype: float64]

The larger this different, the more predictive the feature. Why do these make sense (or not)?

Answer:
In this case the most predictive feature is based on the above factor involving the means of the two classes used for classification: true, false. We see that the most predictive feature has the highest mean difference, and hence can say that  more the mean difference between the classification values , more it shows that the attribute is doing better classification as it is dividing the dataset into more purer sets, thus giving better classification. 
 
 Since, we are using NbGaussian the most important feature we should consider Mean, Standard Deviation.
	Here we see that the most pedictive feature is assigned on the  basis that the highest difference in the mean values is assigned as the most predictive feature because higher the difference - easier it will be to  divide the dataset into pure sets and thus better will be the classification of the dataset. 


iii.	How do these results compare with your results from decision trees, above?
    

<table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td><table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td>
    <table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td>
    
  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>

  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>

    
  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>


We see that both desicion tree classifier and NbGaussian gives “PctKids2Par ”: population of children
with two parents as the most predictive feature with importance of  0.358 and 1.881 respectively.
                        
                        
	•   However, we see that accuracy and presicion for NBGaussian is better than decision Tree. This
    	shows that for NbGaussian most of the true classification predicted are true than Desicion Tree.
	•   Here we also see that, the better the accuracy and  recall are  better the classifier, as we need to know
      	the true classes and its important to know the true prediction of it than just to know the how precisely we predicted the crime rate area.Thsu making NBgauss as better classification here.


# Use LinearSVC 

#### use linear SVC to learn a linear Support Vector Machine model to predict highCrime.

Support vector machines (SVMs) are a set of supervised learning methods used for classification, regression and outliers detection. It uses a technique called the kernel trick to transform your data and then based on these transformations it finds an optimal boundary between the possible outputs.

Why use linearSVC:                                                                                                     
    Linear svm works fine if your dataset is linearly seperable.                                                       
    Linear svm is a parametric model and non-linear svm isn’t for which the complexity grows.                          


In [131]:
clf_svc = LinearSVC()

i.	 What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [133]:
print(" for folds")
accuracy_dt = cross_val_score(clf_svc, dataset1,dataset['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_svc, dataset1,dataset['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_svc, dataset1,dataset['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_svc , dataset1, dataset['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_svc , dataset1, dataset['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_svc , dataset1, dataset['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds
accuracy
[ 0.79        0.865       0.84        0.80904523  0.69346734  0.69346734
  0.81407035  0.8241206   0.81909548  0.81407035]
Precision
[ 0.76774194  0.84507042  0.82517483  0.85950413  0.87209302  0.7962963
  0.84920635  0.95918367  0.80689655  0.87288136]
Recall
[ 0.952  0.96   0.944  0.832  0.6    0.688  0.856  0.752  0.936  0.824]
Prec 0.845405
Recall 0.834400
Accuracy 0.796234


ii.	What are the 10 most predictive features? 

In [135]:
clf_svc.fit(dataset1,dataset['highcrime'])
total_feat = dataset1.columns
imp_feat = clf_svc.coef_[0]
indices = np.argsort(imp_feat)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature most predictive - {} ({})".format(i+1,total_feature[indices[i]],imp_feat[indices[i]]))


Feature Ranking
1 feature most predictive - PersPerOccupHous (1.755126955446997)
2 feature most predictive - RentHighQ (1.066884652098207)
3 feature most predictive - MalePctDivorce (1.065695691001046)
4 feature most predictive - NumUnderPov (1.0515482261273639)
5 feature most predictive - NumStreet (1.0191556627378162)
6 feature most predictive - PctOccupMgmtProf (1.0146708958988186)
7 feature most predictive - population (1.0023013744842268)
8 feature most predictive - agePct12t21 (0.9882733735333677)
9 feature most predictive - HousVacant (0.9365135224585202)
10 feature most predictive - agePct65up (0.8905733237997868)


This can be measured by the absolute feature weights in the model. Why do these make sense (or not)?


Answer : Feature selection is a process by which you automatically search for the best subset of attributes in your dataset.  The best features combination gives us highest accuracy which means less misleading of data. 


iii.	How do these results compare with your results from decision trees, above?

    we see there is a chnage in results this is beacuse:

    we know that the accuracy , precison and recall all are higher for the SVC as compared to 
  	decision tree thus it can be considered better than decison tree. SVC is better as it has more flexibile towards       the penalties like constraints or thresholds and loss functions and are better used to scale larger
  	samples.  

# 3. Regression:

a.	Use LinearRegression to learn a linear model directly predicting the crime rate per capita (ViolentCrimesPerPop).

In [139]:
clf_regr = linear_model.LinearRegression()
clf_regr.fit(dataset1,dataset['ViolentCrimesPerPop'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

i.	Using 10-fold cross-validation, what is the estimated mean-squared-error (MSE) of the model?

In this case MSE is the mean value of the squared deviations of the predictions from the true values.


In [273]:
prediction = clf_regr.predict(dataset1)
mse_scores_regr = cross_val_score(clf_regr, dataset1,dataset['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' , cv=10)
print("fold value")
print(mse_scores_regr)
print("MSE")
mse_scores_regr.mean()

fold value
[-0.02983448 -0.02048526 -0.02735939 -0.02320408 -0.01944074 -0.01262621
 -0.01844842 -0.01222962 -0.02183042 -0.01548108]
MSE


-0.02009396930444533

ii.	What is the MSE on the training set (train on all the data then test on it all)?

When we train and test the same dataset we observer that the MSE value is very low. Which signifies that if the mean squared error is
smaller then the closer we are in finding the line of the best fit.

In [275]:
mean_squared_error(dataset['ViolentCrimesPerPop'], prediction)


0.016516774880307179

iii.	What features are most predictive of a high crime rate? A low crime rate?

When we include the coefficients of regresssion model in order to select and interpret the features we have found out the following points. When all features are on the same scale then the most important features will have the highest coeeficients in the model whereas, features uncorelated with the output variables will have coefficientvalues close to zero.


In [276]:
total_feature = dataset1.columns
important_feature = clf_regr.coef_
#important_feature
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('High Feature Ranking') 
for i in range(10):
    print ("{} Top 10 high crime rate feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))
print(" ")
print ('Low Feature Ranking') 
for i in range(10):
    print ("{} Top 10 low crime rate feature - {} ({})".format(i+1,total_feature[index[i]],important_feature[index[i]]))


High Feature Ranking
1 Top 10 high crime rate feature - PersPerOccupHous (0.635088116498612)
2 Top 10 high crime rate feature - PctHousOwnOcc (0.5681332098865257)
3 Top 10 high crime rate feature - MalePctDivorce (0.4585170486416619)
4 Top 10 high crime rate feature - PctRecImmig8 (0.43251055765291124)
5 Top 10 high crime rate feature - MedRent (0.3727277977113875)
6 Top 10 high crime rate feature - medFamInc (0.28797887494607016)
7 Top 10 high crime rate feature - PctEmploy (0.2484743160847047)
8 Top 10 high crime rate feature - MalePctNevMarr (0.2267279128418035)
9 Top 10 high crime rate feature - PctPersDenseHous (0.2143525641419766)
10 Top 10 high crime rate feature - OwnOccMedVal (0.21287586365359498)
 
Low Feature Ranking
1 Top 10 low crime rate feature - PctPersOwnOccup (-0.6756944788028586)
2 Top 10 low crime rate feature - TotalPctDiv (-0.5619243144146536)
3 Top 10 low crime rate feature - whitePerCap (-0.35101577444078197)
4 Top 10 low crime rate feature - PctKids2Par (-0.322

# Ridge Regression


b.	Now use Ridge regression to reduce the amount of overfitting, using RidgeCV to pick the best alpha from among (10, 1, 0.1, 0.01, and 0.001).
    
    i.	What is the estimated MSE of the model under 10-fold CV?


The least square estimator sometimes might provide a good fit to the training data. However, it will not fit efficiently well to the test data. There is a methodology to leave the requirenment of an unbiased estimator. Ridge regression has a way of placing a particular form of constraint on the parameter that is chosen to minimize the penalized sum of squares.


In [147]:

clf_ridge = linear_model.RidgeCV(alphas = (10, 1, 0.1, 0.01, 0.001))
clf_ridge.fit(dataset1, dataset['ViolentCrimesPerPop'])
mse_ridge = cross_val_score(clf_ridge, dataset1, dataset['ViolentCrimesPerPop'], scoring = 'neg_mean_squared_error' ,cv=10)
print("for folds")
print(mse_ridge)
mse_ridge.mean()

for folds
[-0.0295138  -0.01972075 -0.02795035 -0.02311508 -0.01872138 -0.0120445
 -0.01780565 -0.01212225 -0.02191791 -0.01503854]


-0.019795021348167789

ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [149]:
ridge_predict = clf_ridge.predict(dataset1)
print("MSE after prediction")
mean_squared_error(dataset['ViolentCrimesPerPop'], ridge_predict)

MSE after prediction


0.016763529155169519

iii.	What is the best alpha?



Alpha is known as regularization which is used to set certain penality to reduce the noise over the data. When the apha value is very larger then the regularization effect will dominate the squared loss function and the coefficients will tend to be zero. If the alpha value tends to be zero and the solution implies towards the ordinary least squares then coefficients will tend to be big oscillation. Therefore, alpha should be selected such that it maintains the balance between both.


In [150]:
clf_ridge.alpha_

1.0

iv.	What does this say about the amount of overfitting in linear regression for this problem?
    

MSE of linear Regression : -0.020093969304445319                                                                       
MSE of Ridge Regression:  -0.019795021348168747

MSE is important in regression to determine if the model is over or under-fitted and here we will see that for linear regression the MSE value is very small and hence we can state that there will be very less or close to no overfitting.
Eventhough, Linear regression will  provide results based on some estimates it is modeled on from the training data  coefficients of every variable used hence causing an overfitting problem, as each of the declared coefficient estimates will not state the true patterns and relationship in dataset, but will sample data we have. Ridge can be seeing as, add a penalty term and specify some kind of prior knowledge about how the regression should or should not play out which is found by alpha value.

# c.polynomial features

c.	Now use polynomial features to do quadratic (second-order) polynomial regression.

    i.	What is the estimated MSE of the model under 10-fold CV?


In [221]:
featuresR = dataset1[list(dataset1)].values
labelsR = dataset['ViolentCrimesPerPop'].values
Poly_featuresReg_x = np.array(featuresR)
Poly_labelsReg_y = np.array(labelsR)

clf_poly = PolynomialFeatures(degree=2)
X_poly = clf_poly.fit_transform(Poly_featuresReg_x)


clf_linear = linear_model.LinearRegression()
clf_linear.fit(X_poly, Poly_labelsReg_y)
prediction = clf_poly.fit_transform(Poly_featuresReg_x)
pred = clf_linear.predict(prediction)
print("predicted value")
print(pred)

predicted value
[ 0.06  0.14  1.   ...,  0.12  0.27  0.06]


In [279]:

msr_poly = cross_val_score(clf_linear, Poly_featuresReg_x,  Poly_labelsReg_y, cv=10, scoring='neg_mean_squared_error')
print("fold value")
print(msr_poly)
print("mean value of CV=10")
print(msr_reg.mean())
print("Find the MSE for the dataset")
print(" MSE %f" % cross_val_score(clf_linear , dataset1, dataset['highcrime'] , cv=10, scoring='neg_mean_squared_error' ).mean())

fold value
[-0.02983448 -0.02048526 -0.02735939 -0.02320408 -0.01944074 -0.01262621
 -0.01844842 -0.01222962 -0.02183042 -0.01548108]
mean value of CV=10
-0.0200939693044
Find the MSE for the dataset
 MSE -0.131116


ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [167]:
mean_squared_error(dataset['ViolentCrimesPerPop'], pred)

9.1825945088715955e-29

Does this mean the quadratic model is better than the linear model for this problem?                                   
                                                                                                                        
  MSE of polynomial regression:  mse -0.131116
  MSE of linear regression:  mse -0.020093969304445319
    	 
  The mse values for both quadratic and linear regression is found approximately similar, hence we can say that both will give approximately similar results. Hence, we can prove that the curve generated in both the cases are linear by nature.If the results were better for polynomial the reason could be that the data points are more covered for polynomial curve. 



# 4.	Dirty Data

Repeat the decision tree learning question for the full (non-clean) data set and present the results.

In [252]:
dataset_dirty_final = pd.read_csv("E:\machine learning\project\Crime Prediction Data\communities-crime-full.csv")
dataset_dirty = dataset_dirty_final.copy()
dataset_dirty['highcrime'] = dataset_dirty['ViolentCrimesPerPop'].map(lambda x: True if x > 0.1 else False)

In [254]:
dataset_dirty1 = dataset_dirty.drop(['state','communityname','fold','ViolentCrimesPerPop','county','community','highcrime'], axis=1) 

In [255]:
dataset_dirty

state county community           communityname  fold  population  \
0         8      ?         ?            Lakewoodcity     1        0.19   
1        53      ?         ?             Tukwilacity     1        0.00   
2        24      ?         ?            Aberdeentown     1        0.00   
3        34      5     81440     Willingborotownship     1        0.04   
4        42     95      6096       Bethlehemtownship     1        0.01   
5         6      ?         ?       SouthPasadenacity     1        0.02   
6        44      7     41500             Lincolntown     1        0.01   
7         6      ?         ?               Selmacity     1        0.01   
8        21      ?         ?           Hendersoncity     1        0.03   
9        29      ?         ?             Claytoncity     1        0.01   
10        6      ?         ?            DalyCitycity     1        0.13   
11       36      ?         ?  RockvilleCentrevillage     1        0.02   
12       25     21     44105             Needhamtown     1        0.03   
13       55     87     30075          GrandChutetown     1        0.01   
14        6      ?         ?           DanaPointcity     1        0.04   
15       19    187     91370           FortDodgecity     1        0.03   
16       36      1      1000              Albanycity     1        0.15   
17       34     27     17650        Denvilletownship     1        0.01   
18       18      ?         ?          Valparaisocity     1        0.02   
19       42    129     66376       Rostravertownship     1        0.00   
20        6      ?         ?             Modestocity     1        0.25   
21       12     31         ?        Jacksonvillecity     1        1.00   
22       41      ?         ?        KlamathFallscity     1        0.01   
23       19    193     93926           SiouxCitycity     1        0.11   
24        6      ?         ?              Delanocity     1        0.02   
25        8      ?         ?              Goldencity     1        0.00   
26        6      ?         ?             Gardenacity     1        0.06   
27       39     29     61798           Perrytownship     1        0.01   
28       54      ?         ?             Beckleycity     1        0.01   
29        9      7     22490         EastHamptontown     1        0.00   
...     ...    ...       ...                     ...   ...         ...   
1964     36    119     49121         MountVernoncity    10        0.09   
1965     18     97         ?        Indianapoliscity    10        1.00   
1966      6      ?         ?         RedwoodCitycity    10        0.09   
1967     25     17     76135            Westfordtown    10        0.01   
1968     34     25     62430          RedBankborough    10        0.00   
1969      6      ?         ?        SanFranciscocity    10        1.00   
1970     45      ?         ?               Caycecity    10        0.00   
1971      1      ?         ?             Decaturcity    10        0.06   
1972      6      ?         ?            Redlandscity    10        0.08   
1973     39      ?         ?           Englewoodcity    10        0.00   
1974     37      ?         ?             Tarborotown    10        0.00   
1975     53      ?         ?            Kirklandcity    10        0.05   
1976     25     17      1605           Arlingtontown    10        0.06   
1977     47      ?         ?           Maryvillecity    10        0.01   
1978      9      1      4720              Betheltown    10        0.01   
1979     34     39     61530              Rahwaycity    10        0.02   
1980     25      5     69170             Tauntoncity    10        0.06   
1981      9      9     35650              Hamdentown    10        0.07   
1982     42    133     73168    SpringGardentownship    10        0.00   
1983     13      ?         ?          ForestParkcity    10        0.01   
1984     39     95     48342              Maumeecity    10        0.01   
1985      1      ?         ?       AlexanderCitycity    10        0.01   
1986      

To replace the missing values in the dataset we have used mean of every column. The reason is that when we plot a gaussian distribution graph we get a normal distribution curver i.e. the bell curve which in turn also implies that it is a good approach to replace the missing values by mean.

If the distribution curver would have been skewed then we wouldn't have been able to use mean and hence we would have to take the next approach which would have been median.

However, since we have a normal distribution therefore we use mean.

Clean the dirty dataset and replace the missing values by mean

In [262]:
for i in dataset_dirty1.columns:
    #convert objects will change the object type tp numeric
    dataset_dirty1[[i]] = dataset_dirty1[[i]].convert_objects(convert_numeric=True)
    #fillna will help us replace the missing values in our case we are using median because mean is not normally distributed
    dataset_dirty1[[i]] = dataset_dirty1[[i]].fillna(dataset_dirty1[[i]].mean())

C:\Users\Prajakta\Anaconda3\lib\site-packages\ipykernel\__main__.py:3: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  app.launch_new_instance()


In [263]:
clf_tree_dirty = DecisionTreeClassifier()
without_split_clf_dirty = clf_tree_dirty.fit(dataset_dirty1, dataset_dirty['highcrime'])
prediction = without_split_clf_dirty.predict(dataset_dirty1)
print("Accuracy: %2f" %clf_tree_dirty.score(dataset_dirty1, dataset_dirty['highcrime']))
print("The Recall is :",recall_score( dataset_dirty['highcrime'], prediction, average='weighted'))
print("The Precision is :",precision_score( dataset_dirty['highcrime'], prediction, average='weighted'))  
print ("\n\nThe report:-")
print (metrics.classification_report( dataset_dirty['highcrime'],prediction),"\n")


Accuracy: 1.000000
The Recall is : 1.0
The Precision is : 1.0


The report:-
             precision    recall  f1-score   support

      False       1.00      1.00      1.00       743
       True       1.00      1.00      1.00      1251

avg / total       1.00      1.00      1.00      1994
 



In [264]:
dataset_dirty1

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
5           0.02           0.28          0.06          0.54          1.00   
6           0.01           0.39          0.00          0.98          0.06   
7           0.01           0.74          0.03          0.46          0.20   
8           0.03           0.34          0.20          0.84          0.02   
9           0.01           0.40          0.06          0.87          0.30   
10          0.13           0.71          0.15          0.07          1.00   
11          0.02           0.46          0.08          0.91          0.07   
12          0.03           0.47          0.01          0.96          0.13   
13          0.01           0.44          0.00          0.98          0.04   
14          0.04           0.36          0.01          0.85          0.14   
15          0.03           0.34          0.06          0.93          0.03   
16          0.15           0.31          0.40          0.63          0.14   
17          0.01           0.53          0.01          0.94          0.20   
18          0.02           0.47          0.01          0.97          0.07   
19          0.00           0.41          0.05          0.96          0.01   
20          0.25           0.54          0.05          0.71          0.48   
21          1.00           0.42          0.47          0.59          0.12   
22          0.01           0.34          0.02          0.87          0.07   
23          0.11           0.43          0.04          0.89          0.09   
24          0.02           0.96          0.05          0.00          1.00   
25          0.00           0.33          0.02          0.91          0.16   
26          0.06           0.49          0.46          0.00          1.00   
27          0.01           0.37          0.01          0.99          0.02   
28          0.01           0.27          0.43          0.64          0.08   
29          0.00           0.50          0.01          0.98          0.02   
...          ...            ...           ...           ...           ...   
1964        0.09           0.45          1.00          0.08          0.11   
1965        1.00           0.36          0.44          0.64          0.06   
1966        0.09           0.41          0.07          0.74          0.39   
1967        0.01           0.66          0.00          0.97          0.10   
1968        0.00           0.24          0.51          0.53          0.11   
1969        1.00           0.29          0.21          0.29          1.00   
1970        0.00           0.37          0.38          0.69          0.04   
1971        0.06           0.39          0.32          0.73          0.04   
1972        0.08           0.49          0.07          0.69          0.27   
1973        0.00           0.38          0.03          0.97          0.04   
1974        0.00           0.41          0.62          0.51          0.02   
1975        0.05           0.27          0.03          0.89          0.27   
1976        0.06           0.29          0.03          0.93          0.18   
1977        0.01           0.35          0.06          0.94          0.06   
1978        0.01           0.54          0.02          0.94          0.13   
1979        0.02           0.43          0.39          0.63          0.15   
1980        0.06           0.43          0.04          0.93          0.03   
1981        0.07           0.38          0.17          0.84          0.11   
1982        0.00           0.47          0.02          0.98          0.03   
1983        0.01           0.45          0.37          0.65          0.15   
19

In [178]:
print(" for folds")
accuracy_dt = cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_tree , dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='accuracy' ).mean())

important_feature = clf_tree.feature_importances_
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
print ('Feature Ranking - decision tree dirty data') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

 for folds
accuracy
[ 0.82089552  0.745       0.735       0.75879397  0.74371859  0.7638191
  0.83919598  0.75376884  0.72864322  0.72361809]
Precision
[ 0.84677419  0.8         0.79661017  0.8046875   0.7983871   0.825
  0.85123967  0.83760684  0.77235772  0.78688525]
Recall
[ 0.8015873  0.808      0.76       0.824      0.76       0.784      0.832
  0.808      0.776      0.8      ]
Prec 0.810738
Recall 0.796933
Accuracy 0.763765
Feature Ranking - decision tree dirty data
1 Top 10 high feature - PctKids2Par (0.3586522413947397)
2 Top 10 high feature - racePctWhite (0.08881685598370975)
3 Top 10 high feature - racePctHisp (0.04858535565953847)
4 Top 10 high feature - PctLess9thGrade (0.02263011509341309)
5 Top 10 high feature - PctEmplManu (0.018823516048132382)
6 Top 10 high feature - MedRent (0.01558682991463624)
7 Top 10 high feature - PctHousOwnOcc (0.014165626021920983)
8 Top 10 high feature - OtherPerCap (0.013541829814025207)
9 Top 10 high feature - blackPerCap (0.013365762156192

a.	Are the CV results better or worse? What does this say about the effect of missing values?
 
 The results of dirty data is better as compared to clean data from the table values shown below. Highest weight to validate should be considered of recall and accuracy. Recall(0.796933) and accuracy(0.76160) of dirty data is greater for dirty data. Another reason is in dirty data we have more features added to calculate.

<table style="width:100%">
  <tr>
  
    <th>Clean Data</th>
    <th>Dirty Data</th> 
  </tr>
  <tr>
   
    <td>Prec 0.779517</td>
    <td>Prec 0.810738 </td>
    
    
  </tr>
  <tr>
    <td>Recall 0.782400</td>
    <td>Recall 0.796933</td>
    
  </tr>
  <tr>
    
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.763765</td>
    
  </tr>
  
  
  <tr>
    
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>


# 5.	Extra Credit 

i.	Experiment with two learning methods other than those described above (one can be a non-linear kernel for SVM) for the classification problem, explaining clearly what you did. Show CV results for both the clean and full datasets.

Logistic Regression for clean data

In [193]:

#logistic regression
clf_logic = linear_model.LogisticRegression()

clf_logic.fit(dataset1,dataset['highcrime'])

rs1 = cross_val_score(clf_logic,dataset1, dataset['highcrime'], cv=10)
print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_logic,dataset1, dataset['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_logic,dataset1, dataset['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_logic,dataset1, dataset['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())

important_feature = clf_logic.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('Feature Ranking for logistic regression - High features') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

Cross validation for folds
[ 0.78        0.865       0.885       0.82914573  0.70854271  0.70351759
  0.81407035  0.82914573  0.83417085  0.80904523]
 for folds
accuracy
[ 0.78        0.865       0.885       0.82914573  0.70854271  0.70351759
  0.81407035  0.82914573  0.83417085  0.80904523]
Precision
[ 0.75796178  0.83561644  0.875       0.85826772  0.87640449  0.82352941
  0.84920635  0.96907216  0.82394366  0.85365854]
Recall
[ 0.952  0.976  0.952  0.872  0.624  0.672  0.856  0.752  0.936  0.84 ]
 
 
mean values of accuracy,precision,recall
accuracy
0.805763819095
Precision
0.852266055719
Recall
0.8432
Feature Ranking for logistic regression - High features
1 Top 10 high feature - racepctblack (2.405086901376569)
2 Top 10 high feature - MalePctDivorce (1.4050053029955605)
3 Top 10 high feature - racePctHisp (1.3484991096686791)
4 Top 10 high feature - PctPersDenseHous (1.3083876556037626)
5 Top 10 high feature - PctIlleg (1.089459937105867)
6 Top 10 high feature - TotalPctDiv (1.088

Logistic regression for dirty data

In [197]:
#logistic regression for dirty data
clf_logic_dirty = linear_model.LogisticRegression()

clf_logic_dirty.fit(dataset_dirty1, dataset_dirty['highcrime'])

rs_dirty = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10)
print("For dirty data - Logistic regression")
print(" ")
print(" ")
print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")
important_feature = clf_logic_dirty.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
print(accuracy_dt)
precision_dt = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_logic_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")
important_feature = clf_logic_dirty.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('Feature Ranking for logistic regression(dirty data) - High features') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

For dirty data - Logistic regression
 
 
Cross validation for folds
[ 0.82        0.845       0.845       0.81909548  0.64824121  0.64824121
  0.74874372  0.83919598  0.81909548  0.84422111]
 for folds
accuracy
[ 0.84577114  0.825       0.85        0.82914573  0.81909548  0.83417085
  0.86934673  0.83919598  0.80904523  0.83919598]
Precision
[ 0.88617886  0.8515625   0.88        0.85826772  0.8677686   0.86507937
  0.88976378  0.8907563   0.848       0.872     ]
Recall
[ 0.86507937  0.872       0.88        0.872       0.84        0.872       0.904
  0.848       0.848       0.872     ]
 
 
mean values of accuracy,precision,recall
accuracy
0.835996712418
Precision
0.870937712049
Recall
0.867307936508
 
 
[ 0.84577114  0.825       0.85        0.82914573  0.81909548  0.83417085
  0.86934673  0.83919598  0.80904523  0.83919598]
Precision
[ 0.88617886  0.8515625   0.88        0.85826772  0.8677686   0.86507937
  0.88976378  0.8907563   0.848       0.872     ]
Recall
[ 0.86507937  0.872      

Perceptron for clean data

In [278]:
#perceptron for clean data
clf_perceptron = Perceptron()

clf_perceptron.fit(dataset1,dataset['highcrime'])

rs1 = cross_val_score(clf_perceptron,dataset1, dataset['highcrime'], cv=10)
print("Perceptron calculations:")
print(" ")
print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_perceptron,dataset1, dataset['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_perceptron,dataset1, dataset['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_perceptron,dataset1, dataset['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())

important_feature = clf_perceptron.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('Feature Ranking for Perceptron - High features') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

Perceptron calculations:
 
Cross validation for folds
[ 0.82        0.845       0.845       0.81909548  0.64824121  0.64824121
  0.74874372  0.83919598  0.81909548  0.84422111]
 for folds
accuracy
[ 0.82        0.845       0.845       0.81909548  0.64824121  0.64824121
  0.74874372  0.83919598  0.81909548  0.84422111]
Precision
[ 0.81118881  0.94339623  0.81333333  0.9009009   0.89855072  0.93650794
  0.93103448  0.86614173  0.82962963  0.86153846]
Recall
[ 0.928  0.8    0.976  0.8    0.496  0.472  0.648  0.88   0.896  0.896]
 
 
mean values of accuracy,precision,recall
accuracy
0.787683417085
Precision
0.879222223919
Recall
0.7792
Feature Ranking for Perceptron - High features
1 Top 10 high feature - racepctblack (16.71999999999997)
2 Top 10 high feature - racePctHisp (12.210000000000003)
3 Top 10 high feature - PctPersDenseHous (10.07000000000001)
4 Top 10 high feature - MalePctDivorce (9.340000000000018)
5 Top 10 high feature - PctIlleg (9.22)
6 Top 10 high feature - OtherPerCap (8.

In [199]:
#perceptron for dirty data
clf_perceptron_dirty = Perceptron()

clf_perceptron_dirty.fit(dataset_dirty1, dataset_dirty['highcrime'])

rs_dirty = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10)
print("For dirty data - Perceptron")
print(" ")
print(" ")
print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")#perceptron for dirty data
clf_perceptron_dirty = Perceptron()

clf_perceptron_dirty.fit(dataset_dirty1, dataset_dirty['highcrime'])

rs_dirty = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10)
print("For dirty data - Perceptron")
print(" ")
print(" ")
print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_perceptron_dirty,dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")
important_feature = clf_perceptron_dirty.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('Feature Ranking for logistic regression(dirty data) - High features') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")
important_feature = clf_perceptron_dirty.coef_[0]
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
#indices
print ('Feature Ranking for logistic regression(dirty data) - High features') 
for i in range(10):
    print ("{} Top 10 high feature - {} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

For dirty data - Perceptron
 
 
Cross validation for folds
[ 0.82        0.845       0.845       0.81909548  0.64824121  0.64824121
  0.74874372  0.83919598  0.81909548  0.84422111]
 for folds
accuracy
[ 0.65174129  0.785       0.83        0.8040201   0.73869347  0.74371859
  0.6281407   0.68341709  0.69849246  0.81407035]
Precision
[ 0.64285714  0.76623377  0.90265487  0.80714286  0.71856287  0.71264368
  0.6281407   0.66666667  0.67759563  0.81884058]
Recall
[ 1.     0.944  0.816  0.904  0.96   0.992  1.     0.992  0.992  0.904]
 
 
mean values of accuracy,precision,recall
accuracy
0.737729405735
Precision
0.734133876421
Recall
0.9504
 
 
Feature Ranking for logistic regression(dirty data) - High features
1 Top 10 high feature - racepctblack (21.149999999999945)
2 Top 10 high feature - racePctHisp (10.910000000000009)
3 Top 10 high feature - MalePctDivorce (9.630000000000006)
4 Top 10 high feature - TotalPctDiv (8.88999999999997)
5 Top 10 high feature - HousVacant (8.870000000000038)

## Random forest

In [292]:
#random forest for clean data
clf_Random = RandomForestClassifier(n_jobs=2)
clf_Random.fit(dataset1, dataset['highcrime'])

rs1 = cross_val_score(clf_Random, dataset1, dataset['highcrime'], cv=10)
print(rs1)

print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_Random , dataset1, dataset['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_Random , dataset1, dataset['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_Random , dataset1, dataset['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")


# Build a classification task using 3 informative features

important_feature = clf_Random.feature_importances_
import numpy as np
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
indices
print ('Feature Ranking') 
for i in range(10):
    print ("{} Top 10 high feature -{} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))


[ 0.775       0.85        0.87        0.8241206   0.71859296  0.69346734
  0.80904523  0.79396985  0.75879397  0.82914573]
Cross validation for folds
[ 0.775       0.85        0.87        0.8241206   0.71859296  0.69346734
  0.80904523  0.79396985  0.75879397  0.82914573]
 for folds
accuracy
[ 0.78        0.875       0.815       0.8241206   0.71859296  0.70854271
  0.8040201   0.77889447  0.80904523  0.8241206 ]
Precision
[ 0.79333333  0.88888889  0.86259542  0.84496124  0.81521739  0.84042553
  0.85833333  0.94        0.80555556  0.85483871]
Recall
[ 0.952  0.904  0.896  0.824  0.624  0.64   0.856  0.64   0.896  0.848]
 
 
mean values of accuracy,precision,recall
accuracy
0.793733668342
Precision
0.850414940417
Recall
0.808
 
 
Feature Ranking
1 Top 10 high feature -PctKids2Par (0.10004059180804872)
2 Top 10 high feature -NumUnderPov (0.05067465733698347)
3 Top 10 high feature -MalePctDivorce (0.04922282183945829)
4 Top 10 high feature -PctTeen2Par (0.042052917836362134)
5 Top 10 high

In [293]:
#random forest for dirty data
clf_Random2 = RandomForestClassifier(n_jobs=2)
clf_Random2.fit(dataset_dirty1, dataset_dirty['highcrime'])

rs1 = cross_val_score(clf_Random2, dataset_dirty1, dataset_dirty['highcrime'], cv=10)
print(rs1)

print("Cross validation for folds")
print(rs1)
print(" for folds")
accuracy_dt = cross_val_score(clf_Random2, dataset_dirty1, dataset_dirty['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_Random2, dataset_dirty1, dataset_dirty['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_Random2, dataset_dirty1, dataset_dirty['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)
print(" ")
print(" ")
print("mean values of accuracy,precision,recall")
print("accuracy")
print(accuracy_dt.mean())
print("Precision")
print(precision_dt.mean())
print("Recall")
print(recall_dt.mean())
print(" ")
print(" ")


# Build a classification task using 3 informative features

important_feature = clf_Random.feature_importances_
import numpy as np
indices = np.argsort(important_feature)[::-1]
index = np.argsort(important_feature)
indices
print ('Feature Ranking') 
for i in range(10):
    print ("{} Top 10 high feature -{} ({})".format(i+1,total_feature[indices[i]],important_feature[indices[i]]))

[ 0.83084577  0.86        0.815       0.8040201   0.83919598  0.77889447
  0.83417085  0.84422111  0.80904523  0.76884422]
Cross validation for folds
[ 0.83084577  0.86        0.815       0.8040201   0.83919598  0.77889447
  0.83417085  0.84422111  0.80904523  0.76884422]
 for folds
accuracy
[ 0.82587065  0.825       0.87        0.79899497  0.81407035  0.77386935
  0.83417085  0.84422111  0.8040201   0.79396985]
Precision
[ 0.8852459   0.90598291  0.84251969  0.84        0.87096774  0.82786885
  0.86086957  0.90350877  0.85365854  0.85      ]
Recall
[ 0.83333333  0.848       0.88        0.808       0.832       0.84        0.864
  0.824       0.856       0.856     ]
 
 
mean values of accuracy,precision,recall
accuracy
0.818418722968
Precision
0.864062196079
Recall
0.844133333333
 
 
Feature Ranking
1 Top 10 high feature -PctKids2Par (0.10004059180804872)
2 Top 10 high feature -NumUnderPov (0.05067465733698347)
3 Top 10 high feature -MalePctDivorce (0.04922282183945829)
4 Top 10 high fe

#### Analysis of 5.a

            	random forest:        	logistic regresion:         perceptron
          	 
  	clean data-Accuracy 0.7957  	     Accuracy 0.8057          Accuracy 0.787

               	PctKids2Par (0.100)      racepctblack (2.405)    racepctblack (16.71)
               	NumUnderPov (0.0506)     MalePctDivorce (1.405)  racePctHisp (12.21)

            	 
	dirty data-Accuracy 0.8114      	Accuracy 0.835           Accuracy 0.737

            	PctKids2Par (0.100)     racepctblack (2.332)     racepctblack (21.149)
            	NumUnderPov (0.050)     MalePctDivorce (1.335)   racePctHisp (10.910)
           	 
     
     Accuracies in random forest and logistic regresion for dirty data as compared to the clean data which is explained above but for perceptron accuracy is greater in clean data. 

 	If we check both clean data and dirty data logistic regression is giving better results with accuracies of 80% and 83% in case of logistic regression when compared to other models.

	Logistic regression will give better probability scores and gives efficient implementation than random forest and perceptron. Also in  random forest the model is usually observed to overfit for few datasets with noisy classification which is not in case of logistic regression thus providing better results.

###  Requirement of team 3 ( done as extra credit)

ii.	Devise a method to find the most useful threshold for dividing high crime areas from low crime areas (i.e., discretizing XXX to compute highCrime). Define clearly what you mean by “useful”.

Median of the feature ViolentCrimesPerPop is taken because mean will not  give a normal distribution of data.

In [204]:
# dataset preparation using the new threshold of median - Clean data
dataset_TH = dataset_final.copy()
dataset_TH['highcrime'] = dataset_TH['ViolentCrimesPerPop'].map(lambda x: True if x > dataset_TH['ViolentCrimesPerPop'].mean()  else False)

dataset_threshold1= dataset_TH.copy()
dataset_threshold = dataset_threshold1.drop(['state','communityname','fold','ViolentCrimesPerPop','highcrime'], axis=1)

checklist= dataset_threshold1['highcrime'].value_counts()
true_val = (checklist[1]/(dataset_threshold1['highcrime'].shape))*100
false_val =(checklist[0]/(dataset_threshold1['highcrime'].shape))*100
print("true percetage value %2f " %true_val)
print("false percentage value  %2f "%false_val)

true percetage value 35.524335 
false percentage value  64.475665 


Dataset preparation using the new threshold of median - Dirty data

In [210]:
# dataset preparation using the new threshold of median - Dirty data
dataset_dirty_TH = dataset_dirty_final.copy()
dataset_dirty_TH['highcrime'] = dataset_dirty_TH['ViolentCrimesPerPop'].map(lambda x: True if x > dataset_dirty_TH['ViolentCrimesPerPop'].mean()  else False)

dataset_dirty_threshold1= dataset_dirty_TH.copy()
dataset_dirty_threshold = dataset_dirty_threshold1.drop(['state','communityname','fold','ViolentCrimesPerPop','county','community','highcrime'], axis=1)

checklist= dataset_dirty_threshold1['highcrime'].value_counts()
true_val = (checklist[1]/(dataset_dirty_threshold1['highcrime'].shape))*100
false_val =(checklist[0]/(dataset_dirty_threshold1['highcrime'].shape))*100
print("true percetage value for dirty data %2f " %true_val)
print("false percentage value for dirty data %2f "%false_val)

true percetage value for dirty data 35.506520 
false percentage value for dirty data 64.493480 


Update the missing values with median

In [214]:
for i in dataset_dirty_threshold.columns:
    #convert objects will change the object type tp numeric
    dataset_dirty_threshold[[i]] = dataset_dirty_threshold[[i]].convert_objects(convert_numeric=True)
    #fillna will help us replace the missing values in our case we are using median because mean is not normally distributed
    dataset_dirty_threshold[[i]] = dataset_dirty_threshold[[i]].fillna(dataset_dirty_threshold[[i]].median())

C:\Users\Prajakta\Anaconda3\lib\site-packages\ipykernel\__main__.py:3: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  app.launch_new_instance()


In [215]:
dataset_dirty_threshold

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.19           0.33          0.02          0.90          0.12   
1           0.00           0.16          0.12          0.74          0.45   
2           0.00           0.42          0.49          0.56          0.17   
3           0.04           0.77          1.00          0.08          0.12   
4           0.01           0.55          0.02          0.95          0.09   
5           0.02           0.28          0.06          0.54          1.00   
6           0.01           0.39          0.00          0.98          0.06   
7           0.01           0.74          0.03          0.46          0.20   
8           0.03           0.34          0.20          0.84          0.02   
9           0.01           0.40          0.06          0.87          0.30   
10          0.13           0.71          0.15          0.07          1.00   
11          0.02           0.46          0.08          0.91          0.07   
12          0.03           0.47          0.01          0.96          0.13   
13          0.01           0.44          0.00          0.98          0.04   
14          0.04           0.36          0.01          0.85          0.14   
15          0.03           0.34          0.06          0.93          0.03   
16          0.15           0.31          0.40          0.63          0.14   
17          0.01           0.53          0.01          0.94          0.20   
18          0.02           0.47          0.01          0.97          0.07   
19          0.00           0.41          0.05          0.96          0.01   
20          0.25           0.54          0.05          0.71          0.48   
21          1.00           0.42          0.47          0.59          0.12   
22          0.01           0.34          0.02          0.87          0.07   
23          0.11           0.43          0.04          0.89          0.09   
24          0.02           0.96          0.05          0.00          1.00   
25          0.00           0.33          0.02          0.91          0.16   
26          0.06           0.49          0.46          0.00          1.00   
27          0.01           0.37          0.01          0.99          0.02   
28          0.01           0.27          0.43          0.64          0.08   
29          0.00           0.50          0.01          0.98          0.02   
...          ...            ...           ...           ...           ...   
1964        0.09           0.45          1.00          0.08          0.11   
1965        1.00           0.36          0.44          0.64          0.06   
1966        0.09           0.41          0.07          0.74          0.39   
1967        0.01           0.66          0.00          0.97          0.10   
1968        0.00           0.24          0.51          0.53          0.11   
1969        1.00           0.29          0.21          0.29          1.00   
1970        0.00           0.37          0.38          0.69          0.04   
1971        0.06           0.39          0.32          0.73          0.04   
1972        0.08           0.49          0.07          0.69          0.27   
1973        0.00           0.38          0.03          0.97          0.04   
1974        0.00           0.41          0.62          0.51          0.02   
1975        0.05           0.27          0.03          0.89          0.27   
1976        0.06           0.29          0.03          0.93          0.18   
1977        0.01           0.35          0.06          0.94          0.06   
1978        0.01           0.54          0.02          0.94          0.13   
1979        0.02           0.43          0.39          0.63          0.15   
1980        0.06           0.43          0.04          0.93          0.03   
1981        0.07           0.38          0.17          0.84          0.11   
1982        0.00           0.47          0.02          0.98          0.03   
1983        0.01           0.45          0.37          0.65          0.15   
19

Implementation of GaussianNB with threshold of median for - Clean Data

In [219]:
#GaussuanNB for clean data 
clf_GB_TH= GaussianNB()
print(" for folds GaussianNB")
accuracy_dt = cross_val_score(clf_GB_TH , dataset_threshold,dataset_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_GB_TH, dataset_threshold,dataset_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_GB_TH, dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_GB_TH , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_GB_TH , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_GB_TH , dataset1, dataset['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds GaussianNB
accuracy
[ 0.76        0.87        0.825       0.855       0.85        0.76884422
  0.8040201   0.87939698  0.71212121  0.86868687]
Precision
[ 0.64197531  0.85714286  0.75        0.86206897  0.83606557  0.72727273
  0.74242424  0.84057971  0.57471264  0.82352941]
Recall
[ 0.73239437  0.76056338  0.76056338  0.70422535  0.71830986  0.56338028
  0.69014085  0.81690141  0.71428571  0.8       ]
Prec 0.765577
Recall 0.726076
Accuracy 0.761608


In [223]:
#train Gaussian model for clean data
dataset_threshold3 =  dataset_threshold.copy()
clf_GB_TH.fit(dataset_threshold3,dataset_threshold1['highcrime'])
dataset_threshold3['predict'] = clf_GB_TH.predict(dataset_threshold3)


Most predictive features

In [226]:
predict_true_dataset_TH = dataset_threshold3[dataset_threshold3['predict'] == True]
predict_false_dataset_TH = dataset_threshold3[dataset_threshold3['predict'] == False]

Nb_value_TH = []
for i in range(len(dataset_threshold3.columns)):
    Nb_value_TH.append((abs(predict_true_dataset_TH[[i]].mean() - predict_false_dataset_TH[[i]].mean()) / predict_true_dataset_TH[[i]].std() + predict_false_dataset_TH[[i]].std() ))
sorted(Nb_value_TH, key = itemgetter(0), reverse= True)[:10]

[predict    inf
 dtype: float64, medFamInc    2.153427
 dtype: float64, pctWInvInc    2.118491
 dtype: float64, medIncome    2.041063
 dtype: float64, PctKids2Par    2.027361
 dtype: float64, PctPersOwnOccup    1.836633
 dtype: float64, PctFam2Par    1.826311
 dtype: float64, perCapInc    1.751806
 dtype: float64, PctYoungKids2Par    1.73424
 dtype: float64, PctPopUnderPov    1.696895
 dtype: float64]

Implementation of GaussianNB with threshold of median for - Dirty Data

In [218]:
#GaussuanNB for dirty data 
clf_GB_dirty= GaussianNB()
print(" for folds GaussianNB dirty data")
accuracy_dt = cross_val_score(clf_GB_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_GB_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_GB_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_GB_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_GB_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_GB_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds GaussianNB dirty data
accuracy
[ 0.81        0.775       0.815       0.825       0.805       0.785
  0.81909548  0.73869347  0.79292929  0.76262626]
Precision
[ 0.77966102  0.80952381  0.79310345  0.8         0.84782609  0.75925926
  0.84313725  0.70212766  0.79591837  0.75555556]
Recall
[ 0.64788732  0.47887324  0.64788732  0.67605634  0.54929577  0.57746479
  0.6056338   0.46478873  0.55714286  0.48571429]
Prec 0.788611
Recall 0.569074
Accuracy 0.792834


In [229]:
#train Gaussian model for dirty data
dataset_dirty_threshold3 =  dataset_dirty_threshold.copy()
clf_GB_TH.fit(dataset_dirty_threshold3,dataset_dirty_threshold1['highcrime'])
dataset_dirty_threshold3['predict'] = clf_GB_TH.predict(dataset_dirty_threshold3)


Highest features for dirty data GaussianNB

In [231]:
dirty_true_dataset_TH = dataset_dirty_threshold3[dataset_dirty_threshold3['predict'] == True]
dirty_false_dataset_TH = dataset_dirty_threshold3[dataset_dirty_threshold3['predict'] == False]

Nb_value_TH_dirty = []
for i in range(len(dataset_dirty_threshold3.columns)):
    Nb_value_TH_dirty.append((abs(dirty_true_dataset_TH[[i]].mean() - dirty_false_dataset_TH[[i]].mean()) / dirty_true_dataset_TH[[i]].std() + dirty_false_dataset_TH[[i]].std() ))
sorted(Nb_value_TH_dirty, key = itemgetter(0), reverse= True)[:10]

[predict    inf
 dtype: float64, PctKids2Par    1.653405
 dtype: float64, PctPersOwnOccup    1.587826
 dtype: float64, racePctWhite    1.523689
 dtype: float64, PctFam2Par    1.513799
 dtype: float64, pctWInvInc    1.475654
 dtype: float64, FemalePctDiv    1.444
 dtype: float64, PctYoungKids2Par    1.421686
 dtype: float64, PctIlleg    1.418626
 dtype: float64, PctHousLess3BR    1.418291
 dtype: float64]

Linear SVC implementation with median as threshold

Clean data LinearSVC

In [232]:
clf_svc_m = LinearSVC()

In [235]:
print(" for folds LinearSVC")
accuracy_dt = cross_val_score(clf_svc_m, dataset_threshold,dataset_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_svc_m, dataset_threshold,dataset_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_svc_m, dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print(" ")
print("mean values")
print("Prec %f" % cross_val_score(clf_svc_m , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_svc_m , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_svc_m ,dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds LinearSVC
accuracy
[ 0.705       0.85        0.84        0.855       0.865       0.7638191
  0.79899497  0.90954774  0.76262626  0.81313131]
Precision
[ 0.56122449  0.78082192  0.8         0.86206897  0.90740741  0.72222222
  0.76271186  0.92063492  0.64197531  0.77966102]
Recall
[ 0.77464789  0.8028169   0.73239437  0.70422535  0.69014085  0.54929577
  0.63380282  0.81690141  0.74285714  0.65714286]
 
mean values
Prec 0.773873
Recall 0.710423
Accuracy 0.816312


In [236]:
#most predictive features for LinearSVC -Clean Data
clf_svc_m.fit(dataset_threshold,dataset_threshold1['highcrime'])
total_feat = dataset_threshold.columns
imp_feat = clf_svc_m.coef_[0]
indices = np.argsort(imp_feat)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature most predictive - {} ({})".format(i+1,total_feature[indices[i]],imp_feat[indices[i]]))


Feature Ranking
1 feature most predictive - NumUnderPov (1.7886522354746153)
2 feature most predictive - PctForeignBorn (1.2111145789960944)
3 feature most predictive - PctIlleg (1.1284640843166933)
4 feature most predictive - PctEmploy (1.1277644209550752)
5 feature most predictive - PersPerOccupHous (0.9563936015865955)
6 feature most predictive - medFamInc (0.9357618888285462)
7 feature most predictive - PctNotHSGrad (0.9277731817082466)
8 feature most predictive - MedRent (0.8898530519491443)
9 feature most predictive - racePctHisp (0.8878043958706221)
10 feature most predictive - HousVacant (0.7075377387849259)


Dirty data LinearSVC

In [238]:
clf_svc_m_dirty = LinearSVC()

In [239]:
print(" for folds LinearSVC dirty data")
accuracy_dt = cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print(" ")
print("mean values for Linear SVC dirty data")
print("Prec %f" % cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_svc_m_dirty, dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds LinearSVC dirty data
accuracy
[ 0.87        0.85        0.84        0.835       0.855       0.83
  0.84422111  0.84924623  0.85858586  0.82828283]
Precision
[ 0.85714286  0.80597015  0.76        0.77142857  0.875       0.79365079
  0.78571429  0.82539683  0.828125    0.78125   ]
Recall
[ 0.76056338  0.76056338  0.8028169   0.76056338  0.69014085  0.70422535
  0.77464789  0.73239437  0.75714286  0.71428571]
 
mean values for Linear SVC dirty data
Prec 0.808368
Recall 0.745734
Accuracy 0.846034


In [240]:
#most predictive features for LinearSVC -Dirty Data
clf_svc_m_dirty.fit(dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'])
total_feat = dataset_dirty_threshold.columns
imp_feat = clf_svc_m.coef_[0]
indices = np.argsort(imp_feat)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature most predictive - {} ({})".format(i+1,total_feature[indices[i]],imp_feat[indices[i]]))


Feature Ranking
1 feature most predictive - NumUnderPov (1.7886522354746153)
2 feature most predictive - PctForeignBorn (1.2111145789960944)
3 feature most predictive - PctIlleg (1.1284640843166933)
4 feature most predictive - PctEmploy (1.1277644209550752)
5 feature most predictive - PersPerOccupHous (0.9563936015865955)
6 feature most predictive - medFamInc (0.9357618888285462)
7 feature most predictive - PctNotHSGrad (0.9277731817082466)
8 feature most predictive - MedRent (0.8898530519491443)
9 feature most predictive - racePctHisp (0.8878043958706221)
10 feature most predictive - HousVacant (0.7075377387849259)


Decision tree for new threshold

Clean data

In [244]:
clf_tree_th = DecisionTreeClassifier()
without_split_clf_th = clf_tree_th.fit(dataset_threshold,dataset_threshold1['highcrime'])
prediction = without_split_clf_th.predict(dataset_threshold)

print("Accuracy: %2f" %clf_tree_th.score(dataset_threshold,dataset_threshold1['highcrime']))
print("The Recall is :",recall_score(dataset_threshold1['highcrime'], prediction, average='weighted'))
print("The Precision is :",precision_score(dataset_threshold1['highcrime'], prediction, average='weighted'))  
print ("\n\nThe report:-")
print (metrics.classification_report(dataset_threshold1['highcrime'],prediction),"\n")


Accuracy: 1.000000
The Recall is : 1.0
The Precision is : 1.0


The report:-
             precision    recall  f1-score   support

      False       1.00      1.00      1.00      1285
       True       1.00      1.00      1.00       708

avg / total       1.00      1.00      1.00      1993
 



In [245]:
my_features = dataset_threshold.columns
imp_feature = clf_tree_th.feature_importances_
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature {} ({})".format(i+1,my_features[indices[i]],imp_feature[indices[i]]))


Feature Ranking
1 feature PctIlleg (0.4195417807383689)
2 feature PctKids2Par (0.05485719579817509)
3 feature PctPersDenseHous (0.047494383908839845)
4 feature FemalePctDiv (0.021565576960890103)
5 feature racePctWhite (0.017947592262923563)
6 feature PctTeen2Par (0.015960826435419195)
7 feature HousVacant (0.013421687982099192)
8 feature PctVacMore6Mos (0.012177849262994872)
9 feature PctImmigRec8 (0.011851405957136876)
10 feature LandArea (0.011836540072272292)


In [248]:
print(" for folds decision tree values for clean data")
accuracy_dt = cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_tree_th , dataset_threshold,dataset_threshold1['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds decision tree values for clean data
accuracy
[ 0.675       0.775       0.765       0.745       0.77        0.69849246
  0.77889447  0.82914573  0.67171717  0.8030303 ]
Precision
[ 0.5         0.67948718  0.67532468  0.62857143  0.72463768  0.625
  0.63636364  0.74603175  0.53658537  0.70422535]
Recall
[ 0.70422535  0.73239437  0.73239437  0.63380282  0.69014085  0.52112676
  0.67605634  0.69014085  0.62857143  0.67142857]
Prec 0.656234
Recall 0.672354
Accuracy 0.740093


Dirty Data

In [246]:
clf_tree_th_dirty = DecisionTreeClassifier()
without_split_clf_th_dirty = clf_tree_th_dirty.fit(dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'])
prediction = without_split_clf_th_dirty.predict(dataset_dirty_threshold)

print("Accuracy: %2f" %clf_tree_th_dirty.score(dataset_dirty_threshold,dataset_dirty_threshold1['highcrime']))
print("The Recall is :",recall_score(dataset_dirty_threshold1['highcrime'], prediction, average='weighted'))
print("The Precision is :",precision_score(dataset_dirty_threshold1['highcrime'], prediction, average='weighted'))  
print ("\n\nThe report:-")
print (metrics.classification_report(dataset_dirty_threshold1['highcrime'],prediction),"\n")


Accuracy: 1.000000
The Recall is : 1.0
The Precision is : 1.0


The report:-
             precision    recall  f1-score   support

      False       1.00      1.00      1.00      1286
       True       1.00      1.00      1.00       708

avg / total       1.00      1.00      1.00      1994
 



In [247]:
my_features = dataset_dirty_threshold.columns
imp_feature = clf_tree_th_dirty.feature_importances_
indices = np.argsort(imp_feature)[::-1]
print ('Feature Ranking') 
for i in range(10):
    print ("{} feature {} ({})".format(i+1,my_features[indices[i]],imp_feature[indices[i]]))


Feature Ranking
1 feature PctIlleg (0.4193302497952407)
2 feature PctKids2Par (0.0531538906192262)
3 feature PctPersDenseHous (0.04857628128540851)
4 feature FemalePctDiv (0.021490911891041152)
5 feature pctWInvInc (0.01716749174192164)
6 feature PctVacMore6Mos (0.017092486116965063)
7 feature PctPersOwnOccup (0.015565754158764903)
8 feature racePctWhite (0.01478890354592219)
9 feature HousVacant (0.013417980388456386)
10 feature PctTeen2Par (0.011758851585241389)


In [249]:
print(" for folds")
accuracy_dt = cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10,scoring='accuracy')
print("accuracy")
print(accuracy_dt)
precision_dt = cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'],  cv=10, scoring='precision')
print("Precision")
print(precision_dt)
recall_dt = cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall')
print("Recall")
print(recall_dt)

print("Prec %f" % cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='precision' ).mean())
print ("Recall %f" %cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='recall' ).mean())
print ("Accuracy %f" %cross_val_score(clf_tree_th_dirty , dataset_dirty_threshold,dataset_dirty_threshold1['highcrime'], cv=10, scoring='accuracy' ).mean())

 for folds
accuracy
[ 0.775       0.8         0.755       0.75        0.77        0.77
  0.81407035  0.78894472  0.81313131  0.77272727]
Precision
[ 0.64285714  0.7761194   0.63157895  0.65789474  0.68115942  0.70588235
  0.75        0.70422535  0.69333333  0.68      ]
Recall
[ 0.69014085  0.69014085  0.63380282  0.67605634  0.63380282  0.71830986
  0.74647887  0.70422535  0.77142857  0.72857143]
Prec 0.687518
Recall 0.703461
Accuracy 0.770850


Values after adding threshold median

               Gaussian NB                linear svc                dt tree

     clean data-Accuracy 0.761           Accuracy 0.816          Accuracy 0.740

               medFamInc    2.153      NumUnderPov (1.788)       PctIlleg (0.419)
               pctWInvInc    2.118     PctForeignBorn (1.211)    PctKids2Par (0.054)


    dirty data-Accuracy 0.792          Accuracy 0.846          Accuracy 0.770

              PctKids2Par    1.653     NumUnderPov (1.788)     PctIlleg (0.419)
             PctPersOwnOccup    1.587  PctForeignBorn (1.211)  PctKids2Par (0.053)

with threshold 0.1

<table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td><table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td>
    <table style="width:100%">
  <tr>
  <th>"-----------"</th>
    <th>Decision tree</th>
    <th>GaussianNB</th> 
  </tr>
  <tr>
    <td>Important features</td>
    <td>PctKids2Par 0.358</td>
    <td>PctKids2Par  1.881</td>
  </tr>
  <tr>
    <td> </td>
    <td>racePctWhite 0.088</td>
    <td>pctWInvInc	1.841</td>
  </tr>
  <tr>
    <td> </td>
    <td>Prec 0.779517</td>
    <td>Prec 0.911800</td>
    
  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>

  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>

    
  </tr>
  
  
  <tr>
    <td> </td>
    <td>Recall 0.782400</td>
    <td>Recall 0.692000</td>
    
  </tr>
  
  <tr>
    <td> </td>
    <td>Accuracy 0.728500</td>
    <td>Accuracy 0.76160</td>
    
  </tr>
</table>


From the data above we can clearly state that accuracy of models with threshold as median is greater than accuracy for threshold as 0.1 - decision tree

For Gaussian it is same in both the cases